go/cmd/dolt/commands/blame.go

```diff
@@ -0,0 +1,499 @@
+func maybeRowFromTable(ctx context.Context, t *doltdb.Table, rowPK types.Value) (*row.Row, error) {
+	schema, err := t.GetSchema(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error getting schema from table: %v", err)
+	}
+
+	row, ok, err := t.GetRow(ctx, rowPK.(types.Tuple), schema)
+	if err != nil {
+		return nil, fmt.Errorf("error getting row from table: %v", err)
+	}
+	if !ok {
+		return nil, nil
+	}
+
+	return &row, err
+}
+
+func schemaFromCommit(ctx context.Context, c *doltdb.Commit, tableName string) (schema.Schema, error) {
+	t, err := maybeTableFromCommit(ctx, c, tableName)
+	if err != nil {
+		return nil, fmt.Errorf("error getting table %s from commit: %v", tableName, err)
+	}
+	if t == nil {
+		return nil, fmt.Errorf("no table named %s found in commit", tableName)
+	}
+
+	schema, err := t.GetSchema(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error getting schema from table %s: %v", tableName, err)
+	}
+
+	return schema, nil
+}
+
+func pkColNamesFromCommit(ctx context.Context, c *doltdb.Commit, tableName string) ([]string, error) {
+	schema, err := schemaFromCommit(ctx, c, tableName)
+	if err != nil {
+		return nil, fmt.Errorf("error getting schema for commit: %v", err)
+	}
+	return schema.GetPKCols().GetColumnNames(), nil
+}
+
+// rowChanged returns true if the row identified by `rowPK` changed between the parent-child commit pair
+// represented by `input`
+func rowChanged(ctx context.Context, input blameInput, rowPK types.Value) (bool, error) {
+	parentTable := input.ParentTable
+	childTable := input.Table
+
+	// if the table is in the parent commit but not the child one...something's wrong. bail!
+	if parentTable != nil && childTable == nil {
+		return false, fmt.Errorf("expected to find table with name %v in child commit %s, but didn't", input.TableName, input.Hash)
+	}
+	// if the table is in the child commit but not the parent one, it must be new; return true
+	if childTable != nil && parentTable == nil {
+		return true, nil
+	}
+
+	parentRow, err := maybeRowFromTable(ctx, parentTable, rowPK)
+	if err != nil {
+		return false, fmt.Errorf("error getting row from %s in parent commit %s: %v", input.TableName, input.ParentHash, err)
+	}
+	childRow, err := maybeRowFromTable(ctx, childTable, rowPK)
+	if err != nil {
+		return false, fmt.Errorf("error getting row from %s in child commit %s: %v", input.TableName, input.Hash, err)
+	}
+
+	// if the row is in the parent table but not the child one...something's wrong. bail!
+	if parentRow != nil && childRow == nil {
+		return false, fmt.Errorf("expected to find row with PK %v in table %s in child commit %s, but didn't", rowPK, input.TableName, input.Hash)
+	}
+	// if the row is in the child table but not the parent one, it must be new; return true
+	if childRow != nil && parentRow == nil {
+		return true, nil
+	}
+
+	return !row.AreEqual(*parentRow, *childRow, input.ParentSchema), nil
```

Comment:
Did we get an answer to the question about a new column in `childRow` with a non-nil value returning `false` here?

Reply:
No, I wasn't sure how to update the test. I'm sure I could do so with more effort, but I was thinking...if schema changes count as a change to every row (for now) we can actually avoid this case by checking equality of parent and child schemas.

I added [some code](https://github.com/liquidata-inc/dolt/pull/157/commits/26212307bdf6a25cd2bfca1d4d6bed477f9d0ffc) for this that seems to be behaving well. It probably deserves a BATS test, though.

Does that make sense?